File: class09/python/TCC0901-ReconhecedorAnBn.py
#!/usr/bin/env python3
"""TCC0901-ReconhecedorAnBn.py | Aula 09
Reconhece aⁿbⁿ usando contador — mostra que NÃO é autômato finito.
"""
def reconhece_anbn(palavra: str) -> bool:
    i = 0
    while i < len(palavra) and palavra[i] == 'a':
        i += 1
    n_a = i
    while i < len(palavra) and palavra[i] == 'b':
        i += 1
    n_b = i - n_a
    return i == len(palavra) and n_a == n_b and n_a > 0

def derivar_glc(n: int):
    """Deriva aⁿbⁿ usando GLC: S → aSb | ε"""
    print(f"S", end="")
    for _ in range(n):
        print(f" ⇒ aSb", end="")
    print(f" ⇒ aⁿbⁿ" if n > 0 else " ⇒ ε")

if __name__ == "__main__":
    testes = ["ab", "aabb", "aaabbb", "aab", "abb", "aba"]
    print(f"{'Palavra':<12} {'Resultado'}")
    for w in testes:
        print(f"{w:<12} {'ACEITA' if reconhece_anbn(w) else 'REJEITADA'}")
    print("\nDerivação GLC para a³b³:")
    derivar_glc(3)

File: class09/java/TCC0901-ReconhecedorAnBn.java
// TCC0901-ReconhecedorAnBn.java | Aula 09
// Reconhece aⁿbⁿ usando contador — mostra que NÃO é autômato finito.
// Espelho Java do TCC0901-ReconhecedorAnBn.py (mesmas entradas e saídas).
public class TCC0901ReconhecedorAnBn {

    public static boolean reconheceAnbn(String palavra) {
        int i = 0;
        int n = palavra.length();
        while (i < n && palavra.charAt(i) == 'a') {
            i++;
        }
        int nA = i;
        while (i < n && palavra.charAt(i) == 'b') {
            i++;
        }
        int nB = i - nA;
        return i == n && nA == nB && nA > 0;
    }

    public static void derivarGlc(int n) {
        // Deriva aⁿbⁿ usando GLC: S → aSb | ε
        System.out.print("S");
        for (int k = 0; k < n; k++) {
            System.out.print(" ⇒ aSb");
        }
        System.out.println(n > 0 ? " ⇒ aⁿbⁿ" : " ⇒ ε");
    }

    public static void main(String[] args) {
        String[] testes = {"ab", "aabb", "aaabbb", "aab", "abb", "aba"};
        System.out.printf("%-12s %s%n", "Palavra", "Resultado");
        for (String w : testes) {
            System.out.printf("%-12s %s%n", w, reconheceAnbn(w) ? "ACEITA" : "REJEITADA");
        }
        System.out.println("\nDerivação GLC para a³b³:");
        derivarGlc(3);
    }
}
